oauth2-server/src/main/java/com/clouway/oauth2/IssueNewTokenActivity.java

```diff
@@ -64,27 +70,26 @@ public Response execute(Client client, Request request, DateTime instant) {
     String certKey = randomKey(certificates);
     Pem.Block key = certificates.get(certKey);
 
-
     TokenResponse response = tokens.issueToken(GrantType.AUTHORIZATION_CODE, client, identity, authorization.scopes, instant);
     if (!response.isSuccessful()) {
       return OAuthError.invalidRequest("Token cannot be issued.");
     }
 
     BearerToken accessToken = response.accessToken;
-    String host = request.header("Host");
 
-    String idToken = newIdToken().header(new Header("RS256"))
-            .claims(newClaimSet()
-                    .iss(host)
-                    .aud(client.id)
-                    .sub(identity.id())
-                    .iat(instant.timestamp())
-                    .exp(accessToken.expirationTimestamp())
-                    .certId(certKey)
-                    .customClaims(identity.claims())
-                    .build())
-            .signWith(key)
-            .build();
+    String host = request.header("Host");
+    DefaultJwtBuilder idTokenBuilder = new DefaultJwtBuilder();
+
+    String idToken = idTokenBuilder
+            .setHeaderParam("cid", certKey)//CertificateId - the ID of the certificate that the token was signed with.
+            .setIssuer(host)
+            .setAudience(client.id)
+            .setSubject(identity.id())
+            .setIssuedAt(instant.asDate())
+            .setExpiration(new Date(accessToken.expirationTimestamp()))
+            .claim("identity", identityAsJson(identity).toString())
+            .signWith(SignatureAlgorithm.RS256, parsePem(key))
+            .compact();
 
     return new BearerTokenResponse(accessToken.value, accessToken.ttlSeconds(instant), response.refreshToken, idToken);
```

Comment:
The IssueNewTokenForClientTest doesn't verify that idToken is now returned from the response and it's a valid token.